cupy/core/_routines_statistics.pyx

```diff
@@ -26,55 +32,80 @@ cdef ndarray _ndarray_max(ndarray self, axis, out, dtype, keepdims):
-                                               self.ndim, axis, dtype, order):
-            return cub.device_segmented_reduce(self, cub.CUPY_CUB_MAX, axis,
+                                               self.ndim, reduce_axis, dtype,
+                                               order):
+            return cub.device_segmented_reduce(self, cub.CUPY_CUB_MAX,
+                                               reduce_axis, out_axis,
                                                out=out, keepdims=keepdims)
     return _amax(self, axis=axis, out=out, dtype=dtype, keepdims=keepdims)
 
 
 cdef ndarray _ndarray_min(ndarray self, axis, out, dtype, keepdims):
+    cdef tuple reduce_axis, out_axis
     if cupy.cuda.cub_enabled:
-        if cub.can_use_device_reduce(cub.CUPY_CUB_MIN, self.dtype, self.ndim,
-                                     axis, dtype):
-            return cub.device_reduce(self, cub.CUPY_CUB_MIN, out=out,
-                                     keepdims=keepdims)
+        # if import at the top level, a segfault would happen when import cupy!
+        from cupy.core._kernel import _get_axis
+        reduce_axis, out_axis = _get_axis(axis, self.ndim)
+        if cub.can_use_device_reduce(cub.CUPY_CUB_MIN, self.dtype, out_axis,
+                                     dtype):
+            return cub.device_reduce(self, cub.CUPY_CUB_MIN, out_axis,
+                                     out=out, keepdims=keepdims)
+
         if self.flags.c_contiguous:
             order = 'C'
         elif self.flags.f_contiguous:
             order = 'F'
         else:
             order = None
         if cub.can_use_device_segmented_reduce(cub.CUPY_CUB_MIN, self.dtype,
-                                               self.ndim, axis, dtype, order):
-            return cub.device_segmented_reduce(self, cub.CUPY_CUB_MIN, axis,
+                                               self.ndim, reduce_axis, dtype,
+                                               order):
+            return cub.device_segmented_reduce(self, cub.CUPY_CUB_MIN,
+                                               reduce_axis, out_axis,
                                                out=out, keepdims=keepdims)
     return _amin(self, axis=axis, out=out, dtype=dtype, keepdims=keepdims)
 
 
 # TODO(leofang): this signature is incompatible with NumPy!
 cdef ndarray _ndarray_argmax(ndarray self, axis, out, dtype, keepdims):
+    cdef tuple reduce_axis, out_axis
     if cupy.cuda.cub_enabled:
+        # if import at the top level, a segfault would happen when import cupy!
+        from cupy.core._kernel import _get_axis
+
         # Note that the NumPy signature of argmax only has axis and out, so we
         # need to disable the rest. Moreover, to be compatible with NumPy, axis
         # can only be None or integers
+        if sequence.PySequence_Check(axis):
+            raise TypeError(
+                "'tuple' object cannot be interpreted as an integer")
```

Comment:
Notes:

1. This is to match the NumPy behavior, which throws an TypeError in such cases.
2. However, this is inconsistent with current (incompatible) CuPy reduction behavior (#2595). This PR will forbid a tuple axis from running, instead of silently falling back to old reduction routines.

Reply:
By the way, if you want to raise here, you probably need `except? -1` in the function declaration?

Reply:
Right. I am refactoring things a bit more now to simplify things and will try to keep the existing CuPy behavior here

Reply:
I no longer try to raise an exception from this file in the refactored code. Please see f8a3db2.